pygem/cad/ffd.py

```diff
@@ -0,0 +1,435 @@
+        return comp_curve
+
+    def _enrich_curve_knots(self, bsp_curve):
+        """
+        Takes a Geom_BSplineCurve and adds self.t_knots_to_add poles to it.
+
+        :bsp_curve Geom_BSplineCurve to be enriched
+        """
+        if not isinstance(bsp_curve, Geom_BSplineCurve):
+            raise TypeError("bsp_curve must be a Geom_BSplineCurve")
+        # number of knots is enriched here, if required, to
+        # enhance precision
+        # start parameter of composite curve
+        first_param = bsp_curve.FirstParameter()
+        # end parameter of composite curve
+        last_param = bsp_curve.LastParameter()
+        for i in range(self.t_knots_to_add):
+            bsp_curve.InsertKnot(first_param+ \
+                           i*(last_param-first_param)/self.t_knots_to_add, 1, \
+                           self.tolerance)
+
+
+    def _enrich_surface_knots(self, bsp_surface):
+        """
+        Takes a Geom_Bspline_Surface and adds self.u_knots_to_add
+        and self.v_knots_to_add knots to it in u and v direction respectively.
+
+        :bsp_surface Geom_Bspline_Surface to be enriched
+        """
+        if not isinstance(bsp_surface, Geom_BSplineSurface):
+            raise TypeError("bsp_surface must be a Geom_BSplineSurface")
+        # we will add the prescribed amount of nodes
+        # both along u and v parametric directions
+
+        # bounds (in surface parametric space) of the surface
+        bounds = bsp_surface.Bounds()
+
+        for i in range(self.u_knots_to_add):
+            bsp_surface.InsertUKnot(bounds[0]+ \
+                i*(bounds[1]-bounds[0])/self.u_knots_to_add, 1, self.tolerance)
+        for i in range(self.v_knots_to_add):
+            bsp_surface.InsertVKnot(bounds[2]+ \
+                i*(bounds[3]-bounds[2])/self.v_knots_to_add, 1, self.tolerance)
+
+    def _deform_bspline_curve(self, bsp_curve):
+        """
+        Takes a Geom_Bspline_Curve and deforms it through FFD.
+
+        :bsp_curve Geom_Bspline_Curve to be deformed
+        """
+        if not isinstance(bsp_curve, Geom_BSplineCurve):
+            raise TypeError("bsp_curve must be a Geom_BSplineCurve")
+        # we first extract the poles of the curve
+        # poles number
+        n_poles = bsp_curve.NbPoles()
+        # array containing the poles coordinates
+        poles_coordinates = np.zeros(shape=(n_poles, 3))
+        # cycle over the poles to get their coordinates
+
+        for pole_id in range(n_poles):
+            # gp_Pnt corresponding to the pole
+            pole = bsp_curve.Pole(pole_id + 1)
+            # coordinates are added to array
+            poles_coordinates[pole_id, :] = [pole.X(), pole.Y(), pole.Z()]
+
+
+        # the new poles positions are computed through FFD
+        new_pts = super().__call__(poles_coordinates)
+
+        # the Bspline curve is now looped again to
+        # set the poles positions  to new_points
+        i = 0
+        for pole in range(n_poles):
```

Comment:
you can avoid `i=0` and `i+=1` by just using the pole variable: `i` and `pole` are the same.